File: scripts/short_read_chopper.R
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(optparse))
suppressPackageStartupMessages(library(R.utils))

option_list = list(
  make_option(c("--tsv"), type="character", default=NULL, 
              help="path to tsv with simulated mRNAs", metavar="character"),
  make_option(c("--insert_size"), type="character", default="200,300", 
              help="library insert size range, e.g. 200,300", metavar="character"),
  make_option(c("--read_length"), type="numeric", default=100, 
              help="length of each individual read", metavar="numeric"),
  make_option(c("-t", "--threads"), type="numeric", default=1, 
              help="Number of threads", metavar="numeric"),
  make_option(c("--seq_depth"), type="numeric", default=20000000, 
              help="sequencing depth", metavar="numeric"),
  make_option(c("--tpm_lower_limit"), type="numeric", default=5, 
              help="TPM lower limit", metavar="numeric"),
  make_option(c("--tpm_upper_limit"), type="numeric", default=200, 
              help="TPM upper limit", metavar="numeric"),
  make_option(c("-o", "--dir_out"), type="character", default=".", 
              help="dir out to create temp files", metavar="character"),
  make_option(c("--seed"), type="numeric", default=NULL,
              help="seed for randomization"),
  make_option(c("--fragments"), type="character", default='no', 
              help="to export ground truth", metavar="character"),
  make_option(c("--no_sizeselection"), action="store_true", default=FALSE,
              help="If specified, do not filter fragments by size (keep everything)"),
  make_option(c("--no_fragmentation"), action="store_true", default=FALSE,
              help="If specified, do not fragment; keep full molecule length")
)

#Parse input
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

if (is.null(opt$tsv)){
  stop("path to tsv with simulated mRNAs is missing", call.=FALSE)
}

if (!is.null(opt$seed)){
  set.seed(opt$seed)
}

setDTthreads(threads = opt$threads)
insert_size <- opt$insert_size
insert_size_split <- strsplit(insert_size, ",")[[1]]
insert_size <- c(as.numeric(insert_size_split[1]), as.numeric(insert_size_split[2]))

file <- opt$tsv

#File importing and formating
file_importer <- function(file){
  full_reads <- data.table::fread(file=file, sep = '\t', header = T, stringsAsFactors = FALSE)
  # Ensure transcript_id matches the row index logic used later
  full_reads$transcript_id <- as.numeric(1:nrow(full_reads))
  full_reads$sequence_length <- nchar(full_reads$full_molecule_sequence)
  return(full_reads)
}

#Chopper function
get_reads <- function(lengths, eta_val = 200, insert_size, transcript_id, no_fragmentation=FALSE, no_sizeselection=FALSE) {
  
  if (no_fragmentation) {
    # return a single fragment from 1 to the end of the molecule
    fragments <- data.frame(
      transcript_id = transcript_id,
      read_start = 1,
      read_end = lengths,
      length = lengths, # length is the full length
      size_selection = 'passed_size_selection' 
    )
    
    if (!no_sizeselection) {
      if (fragments$length < insert_size[1] || fragments$length > insert_size[2]) {
        fragments$size_selection <- 'filtered_out'
      }
    }
    
    
    fragments$length <- NULL 
    
    return(fragments)
  }
  
  eta_val <- mean(c(insert_size[1],insert_size[2]))
  
  deltas <- log10(lengths)
  ns_minus_1 <- pmax(round(lengths / eta_val / gamma(1 / deltas + 1)) - 1, 0)
  xis <- lapply(ns_minus_1, function(n) { diff(sort(c(runif(n), 0, 1))) })
  xis_transformed <- mapply(function(x, d) { x^(1 / d) }, xis, deltas, SIMPLIFY = FALSE)
  delta_is <- mapply(function(len, x_t) {
    round(len * x_t / sum(x_t))
  }, lengths, xis_transformed, SIMPLIFY = FALSE)
  
  # Get all the start and end points of the fragments
  starts <- lapply(delta_is, function(d) {
    if (length(d) > 1) {
      c(sample(min(insert_size[1], d[1]), 1), cumsum(d[1:(length(d) - 1)]))
    } else {
      sample(min(insert_size[1], d), 1)
    }
  })
  
  ends <- lapply(delta_is, function(d) {
    if (length(d) > 1) {
      c(cumsum(d[1:(length(d) - 1)]), sum(d) - sample(min(insert_size[1], sum(d) - d[length(d)]), 1))
    } else {
      d
    }
  })
  
  fragments <- data.frame(
    transcript_id = rep(transcript_id, length(delta_is[[1]])),
    read_start = unlist(starts),
    read_end = unlist(ends)
  )
  
  # Create first row from 1 to the first read_start
  first_row <- data.frame(
    transcript_id = fragments$transcript_id[1],
    read_start = 1,
    read_end = fragments$read_start[1]
  )
  
  # Create last row from the last read_end to the value of lengths
  last_row <- data.frame(
    transcript_id = fragments$transcript_id[nrow(fragments)],
    read_start = fragments$read_end[nrow(fragments)],
    read_end = lengths
  )
  
  # Combine all together
  fragments <- rbind(first_row, fragments, last_row)
  fragments$length <- fragments$read_end - fragments$read_start
  
  # Filter by insert size
  fragments$size_selection <- 'filtered_out'
  
  
  if (no_sizeselection) {
    # ff the flag is set, everything passes the selection
  } else {
    fragments$size_selection <- 'passed_size_selection'
    fragments[fragments$length >= insert_size[1] & fragments$length <= insert_size[2], 'size_selection'] <- 'passed_size_selection'
  }
  
  fragments$length <- NULL  # Optional: drop if not needed
  
  return(fragments)
}


random_string_gen <- function(n = 5000) {
  a <- do.call(paste0, replicate(5, sample(LETTERS, n, TRUE), FALSE))
  paste0(a, sprintf("%04d", sample(9999, n, TRUE)), sample(LETTERS, n, TRUE))
}

full_reads <- file_importer(file)
#Get gene that was simulated
gene_id <- sub(".*/(ENSG[0-9]+(?:_background)?)(?:\\.tsv\\.gz)$", "\\1", file)

list_with_all_fragments <- lapply(seq_len(nrow(full_reads)), function(i) {
  get_reads(lengths = full_reads$sequence_length[i],
            insert_size = insert_size,
            transcript_id = full_reads$transcript_id[i],
            # Pass the new options from the opt object
            no_fragmentation = opt$no_fragmentation,
            no_sizeselection = opt$no_sizeselection)
})

reads_list <- dplyr::bind_rows(list_with_all_fragments,.id = 'transcript_id')

#Keep all fragments before filtering by their length
reads_list_all_fragments <- reads_list
#Keep fragments that are long enough
reads_list <- reads_list[reads_list$size_selection=='passed_size_selection',]


if (nrow(reads_list)>10){ #Some genes have really short mRNAs and will get filtered out during size selection
  reads_list$transcript_id <- as.numeric(reads_list$transcript_id)
  #Assign a TPM and sample the number of reads (rows) accordingly
  gene_length <- mean(full_reads$sequence_length)/1000
  gene_tpm <- runif(n=1, min=as.integer(opt$tpm_lower_limit), max=as.integer(opt$tpm_upper_limit))
  seq_depth <- opt$seq_depth/10^6
  
  #Get the number of fragments for the gene. 
  reads_to_get <- gene_length*gene_tpm*seq_depth
  
  reads_list$read_start <- as.integer(reads_list$read_start)
  reads_list$read_end <- as.integer(reads_list$read_end)
  
  if (opt$fragments=='with_ground_truth'){
    #Get a ground truth that belongs to fragmented and size selected fragments (but not mapped)
    all_positions_selected <- unlist(mapply(seq, reads_list$read_start, reads_list$read_end))
    freq_table_selected <- as.data.table(table(all_positions_selected))
    setnames(freq_table_selected, c("position", "frequency"))
    freq_table_selected[, position := as.integer(as.character(position))]
    
    dir_ground_truth <- paste(opt$dir_out,'/ground_truth_after_size_selection',sep = '')
    dir.create(dir_ground_truth, showWarnings = FALSE, recursive = TRUE)
    path_for_file <- paste(dir_ground_truth,'/',gene_id,'.tsv.gz',sep = '')
    fwrite(freq_table_selected,path_for_file,sep = '\t',col.names = T,row.names = F,quote = F,compress = 'gzip')
  }
  
  
  reads_list_collapsed <- reads_list %>%
    dplyr::mutate(read_coordinate = paste0(read_start, "-", read_end)) %>%
    dplyr::group_by(transcript_id) %>%
    dplyr::summarise(read_coordinates = paste(read_coordinate, collapse = ","), .groups = "drop")
  
  #### Get ground truth for all the fragments before fragmentation and size selection
  if (opt$fragments=='with_ground_truth'){
    
    # Avoid division by zero if all fragments pass size selection (which happens with no_sizeselection)
    n_passed <- nrow(reads_list_all_fragments[reads_list_all_fragments$size_selection=='passed_size_selection',])
    if(n_passed == 0) {
      ratio_total_passed <- 0
    } else {
      ratio_total_passed <- nrow(reads_list_all_fragments)/n_passed
    }
    
    #Sample the total reads based on the ratio and the assigned gene TPM
    all_fragments_to_get <- reads_to_get
    
    #Sample multiple times to smooth the distribution
    all_sampling <- data.frame()
    for (j in 1:100){
      if (nrow(reads_list_all_fragments)>=all_fragments_to_get){
        reads_list_all_fragments_sampled <- dplyr::sample_n(reads_list_all_fragments,size=all_fragments_to_get,replace = F) #No replacement
      }else{
        sampled_reads <- dplyr::sample_n(reads_list_all_fragments,size=all_fragments_to_get-nrow(reads_list_all_fragments),replace = T) #With replacement for the reads that are missing
        reads_list_all_fragments_sampled <- rbind(reads_list_all_fragments,sampled_reads)
      }
      
      #Get ground truth for the number of mRNAs that are going into this -> used for edge effect correction
      all_positions <- unlist(mapply(seq, reads_list_all_fragments_sampled$read_start, reads_list_all_fragments_sampled$read_end))
      freq_table <- as.data.table(table(all_positions))
      setnames(freq_table, c("position", "frequency"))
      freq_table[, position := as.integer(as.character(position))]
      all_sampling <- rbind(all_sampling,freq_table)
    }
    final_freqs <- all_sampling %>% dplyr::group_by(position) %>% dplyr::reframe(frequency=mean(frequency))
    final_freqs <- final_freqs[-nrow(final_freqs),]#Remove last nucleotide which will always be noisy
    
    dir_ground_truth <- paste(opt$dir_out,'/ground_truth_pre_size_selection',sep = '')
    dir.create(dir_ground_truth, showWarnings = FALSE, recursive = TRUE)
    path_for_file <- paste(dir_ground_truth,'/', gene_id,'.tsv.gz',sep = '')
    fwrite(final_freqs,path_for_file,sep = '\t',col.names = T,row.names = F,quote = F,compress = 'gzip')
  }
  
  temp_dir <- paste(opt$dir_out,'/temp/mRNAs_with_fragments',sep = '')
  dir.create(temp_dir, showWarnings = FALSE, recursive = TRUE)
  path_for_file <- paste(temp_dir,'/',gene_id,'_fragments.tsv',sep = '')
  
  fwrite(reads_list_collapsed, path_for_file, sep = '\t', col.names = T, row.names = F, quote = F)
  
}

File: scripts/short_read_chopper.py
#!/usr/bin/env python3
"""
this was written to be able to chop acorss seq methods, such that:

tree output modes (--mode) hardcoded into the fastq generators that call thiss cript:
  normal  – used by fastq_generator_short_read.py
            Collapsed output (transcript_id → "start-end,start-end,..."), writes TPM,
            optionally writes coverage ground-truth files.
  proseq  – used by fastq_generator_proseq.py
            Per-row output (transcript_id, molecule_id, read_start, read_end).
  ttseq   – used by mRNA_to_reads_ttseq.py
            Collapsed output, no TPM, drops fragments shorter than 10 nt.

"""

import argparse
import hashlib
import os
import sys

import numpy as np
import pandas as pd
from scipy.special import gamma as sp_gamma



def _apply_size_selection(lengths, ins_min, ins_max, sizeselectiontype, rng):
    """return a boolean to know which fragments pass size selection."""
    if sizeselectiontype == "none":
        return lengths >= 1
    elif sizeselectiontype == "probabilistic":
        p_left  = 1.0 / (1.0 + np.exp(-0.1 * (lengths - ins_min)))
        p_right = 1.0 / (1.0 + np.exp( 0.1 * (lengths - ins_max)))
        rolls   = rng.uniform(0.0, 1.0, len(lengths))
        return rolls <= (p_left * p_right)
    else:  # hardcut
        return (lengths >= ins_min) & (lengths <= ins_max)


def _fragment_molecule_raw(seq_len, ins_min, ins_max, rng):
    eta   = (ins_min + ins_max) / 2.0
    delta = np.log10(max(seq_len, 2))          # log10 must be > 0

    #number of internal cut-points
    n_cuts = max(0, round(seq_len / eta / sp_gamma(1.0 / delta + 1)) - 1)

    if n_cuts > 0:
        pts  = np.sort(rng.uniform(0.0, 1.0, n_cuts))
        xis  = np.diff(np.concatenate([[0.0], pts, [1.0]]))
    else:
        xis = np.array([1.0])

    xis_t = xis ** (1.0 / delta)
    s     = xis_t.sum()
    if s <= 0:
        xis_t = np.ones_like(xis)
        s     = float(len(xis))

    d       = np.round(seq_len * xis_t / s).astype(int)
    d       = np.maximum(d, 1)                #I don't want zero-size fragments
    n       = len(d)
    cumsums = np.cumsum(d)                    #length n


    if n > 1:
        s0_max     = max(1, min(ins_min, int(d[0])))
        random_s   = int(rng.integers(1, s0_max + 1))

        last_gap   = max(1, int(cumsums[-1]) - int(d[-1]))
        last_max   = max(1, min(ins_min, last_gap))
        random_off = int(rng.integers(1, last_max + 1))

        #starts: [random_s, cumsum[0], cumsum[1], ..., cumsum[n-2]]
        int_starts = np.concatenate([[random_s], cumsums[:-1]])
        #ends:   [cumsum[0], ..., cumsum[n-2], sum(d) - random_off]
        int_ends   = np.concatenate([cumsums[:-1], [int(cumsums[-1]) - random_off]])
    else:
        s0_max     = max(1, min(ins_min, int(d[0])))
        random_s   = int(rng.integers(1, s0_max + 1))
        int_starts = np.array([random_s])
        int_ends   = np.array([int(d[0])])

    all_starts = np.concatenate([[1],                int_starts, [int(int_ends[-1])]])
    all_ends   = np.concatenate([[int(int_starts[0])], int_ends,   [seq_len]])

    #filter out zero length entries
    keep       = all_ends > all_starts
    return all_starts[keep], all_ends[keep]


def fragment_molecule(seq_len, ins_min, ins_max, rng,
                      no_fragmentation=False, sizeselectiontype="probabilistic"):
    """
    fragment single molecules and size select
    """
    if seq_len < 1:
        empty = np.array([], dtype=int)
        return empty, empty, empty, empty

    if no_fragmentation:
        s = np.array([1],       dtype=int)
        e = np.array([seq_len], dtype=int)
        lengths = e - s
        passed  = _apply_size_selection(lengths, ins_min, ins_max, sizeselectiontype, rng)
        return s[passed], e[passed], s, e

    raw_s, raw_e = _fragment_molecule_raw(seq_len, ins_min, ins_max, rng)
    lengths      = raw_e - raw_s
    passed       = _apply_size_selection(lengths, ins_min, ins_max, sizeselectiontype, rng)
    return raw_s[passed], raw_e[passed], raw_s, raw_e

def compute_coverage(starts, ends):
    if len(starts) == 0:
        return np.array([], dtype=int), np.array([], dtype=int)
    max_pos = int(ends.max())
    cov = np.zeros(max_pos + 2, dtype=np.int64)
    np.add.at(cov, starts.astype(int),     1)
    np.add.at(cov, ends.astype(int),      -1)
    cov = np.cumsum(cov)[:max_pos]
    pos = np.arange(1, max_pos + 1)
    mask = cov > 0
    return pos[mask], cov[mask]




def main():
    parser = argparse.ArgumentParser(
        description="Short-read chopper – Python port of the three R chopper scripts"
    )
    parser.add_argument("--tsv",             required=True,
                        help="Input TSV.gz with simulated mRNAs (must have sequence_length column)")
    parser.add_argument("--mode",            required=True,
                        choices=["normal", "proseq", "ttseq"],
                        help="Output mode: normal | proseq | ttseq")
    parser.add_argument("--insert_size",     default="200,300",
                        help="Insert-size range min,max  e.g. 200,300")
    parser.add_argument("--read_length",     type=int, default=100,
                        help="Read length (kept for CLI compatibility; not used in fragmentation)")
    parser.add_argument("--threads",         type=int, default=1,
                        help="Threads (kept for CLI compatibility)")
    parser.add_argument("--seq_depth",       type=int, default=20_000_000,
                        help="Total library sequencing depth (used for TPM in normal mode)")
    parser.add_argument("--tpm_lower_limit", type=int, default=5)
    parser.add_argument("--tpm_upper_limit", type=int, default=200)
    parser.add_argument("-o", "--dir_out",   default=".",
                        help="Output directory (same -o as fastq generators)")
    parser.add_argument("--seed",            type=int, default=None)
    parser.add_argument("--fragments",       type=str, default="no",
                        help="Pass 'with_ground_truth' to write coverage files (normal mode only)")
    parser.add_argument("--sizeselectiontype", default="probabilistic",
                        choices=["none", "hardcut", "probabilistic"])
    parser.add_argument("--no_fragmentation", action="store_true",
                        help="Skip fragmentation; treat each molecule as a single fragment")

    args = parser.parse_args()

    ins_min, ins_max = map(int, args.insert_size.split(","))


    gene_id_raw = os.path.basename(args.tsv)
    gene_id     = gene_id_raw
    for ext in (".tsv.gz", ".gz", ".tsv"):
        if gene_id.endswith(ext):
            gene_id = gene_id[: -len(ext)]
            break

    gene_id_for_hash = gene_id.split("_")[0]

    rng = np.random.default_rng()
    if args.seed is not None:
        h           = hashlib.sha256(gene_id_for_hash.encode("utf-8"))
        gene_hash   = int(h.hexdigest(), 16) % (2 ** 32)
        unique_seed = (args.seed + gene_hash) % (2 ** 32)
        rng         = np.random.default_rng(unique_seed)


    if args.mode == "proseq":
        base_part  = gene_id.split("_")[0]
        out_gene_id = (base_part + "_background") if "_background" in gene_id else base_part
    else:
        out_gene_id = gene_id

    try:
        df = pd.read_csv(args.tsv, sep='\t')
        
        if df.empty:
            print(f"Warning: {args.tsv} contains no molecules (likely 100% background). Exiting gracefully.")
            sys.exit(0)
            
        if 'sequence_length' not in df.columns:
            raise ValueError(f"'sequence_length' column not found in {args.tsv}")

    #this is here to account for cases were bkg 1 and nascent mRNA dfs are empty
    except pd.errors.EmptyDataError:
        print(f"Warning: {args.tsv} is completely empty. Ignore this warning if bkg 1 was set")
        sys.exit(0)

    if "full_molecule_sequence" in df.columns:
        df = df.drop(columns=["full_molecule_sequence"])
        
    df["transcript_id"] = range(1, len(df) + 1)
    
    if len(df) <= 5:
        return  


    sel_starts_list = []
    sel_ends_list   = []
    all_starts_list = []
    all_ends_list   = []
    tid_list        = []

    need_pre_cov = (args.mode == "normal" and args.fragments == "with_ground_truth")

    for row in df.itertuples(index=False):
        seq_len = int(row.sequence_length)
        tid     = int(row.transcript_id)

        s_sel, e_sel, s_all, e_all = fragment_molecule(
            seq_len, ins_min, ins_max, rng,
            no_fragmentation=args.no_fragmentation,
            sizeselectiontype=args.sizeselectiontype,
        )

        if len(s_sel) > 0:
            sel_starts_list.append(s_sel)
            sel_ends_list.append(e_sel)
            tid_list.append(np.full(len(s_sel), tid, dtype=int))

        if need_pre_cov and len(s_all) > 0:
            all_starts_list.append(s_all)
            all_ends_list.append(e_all)

    if not sel_starts_list:
        return  # nothing passed size selection

    sel_starts = np.concatenate(sel_starts_list)
    sel_ends   = np.concatenate(sel_ends_list)
    tids       = np.concatenate(tid_list)

    #TTseq: drop fragments shorter than 10nt
    if args.mode == "ttseq":
        keep = (sel_ends - sel_starts) > 10
        sel_starts, sel_ends, tids = sel_starts[keep], sel_ends[keep], tids[keep]
        if len(sel_starts) == 0:
            return

    if len(sel_starts) <= 10:
        return

    dir_out   = args.dir_out.rstrip("/")
    frags_dir = os.path.join(dir_out, "temp", "mRNAs_with_fragments")
    temp_dir  = os.path.join(dir_out, "temp")
    os.makedirs(frags_dir, exist_ok=True)
    os.makedirs(temp_dir,  exist_ok=True)
    out_path = os.path.join(frags_dir, f"{out_gene_id}_fragments.tsv")

    if args.mode == "normal":
        gene_length_kb = float(df["sequence_length"].mean()) / 1000.0
        gene_tpm       = float(rng.uniform(args.tpm_lower_limit, args.tpm_upper_limit))
        seq_depth_m    = args.seq_depth / 1e6
        reads_to_get   = gene_length_kb * gene_tpm * seq_depth_m

        #TPM temp file (may be overwritten by fastq_generator_short_read.py – that's ok)
        tpm_path = os.path.join(temp_dir, f"temp_{out_gene_id}_tpm.tsv")
        pd.DataFrame([{"gene_id": out_gene_id, "tpm": gene_tpm}]).to_csv(
            tpm_path, sep="\t", index=False
        )

        #ground truth if specified
        if args.fragments == "with_ground_truth":
            #oost-size-selection coverage
            pos_sel, cov_sel = compute_coverage(sel_starts, sel_ends)
            if len(pos_sel) > 0:
                gt_sel_dir = os.path.join(dir_out, "ground_truth_after_size_selection")
                os.makedirs(gt_sel_dir, exist_ok=True)
                pd.DataFrame({"position": pos_sel, "frequency": cov_sel}).to_csv(
                    os.path.join(gt_sel_dir, f"{out_gene_id}.tsv.gz"),
                    sep="\t", index=False
                )

            #pre-size-selection coverage (scaled to target depth)
            if all_starts_list:
                all_s = np.concatenate(all_starts_list)
                all_e = np.concatenate(all_ends_list)
                n_passed = len(sel_starts)
                scaling  = (reads_to_get / n_passed) if n_passed > 0 else 0.0
                pos_pre, cov_pre = compute_coverage(all_s, all_e)
                if len(pos_pre) > 0 and scaling > 0:
                    gt_pre_dir = os.path.join(dir_out, "ground_truth_pre_size_selection")
                    os.makedirs(gt_pre_dir, exist_ok=True)
                    pd.DataFrame({
                        "position":  pos_pre,
                        "frequency": (cov_pre * scaling).round().astype(int),
                    }).to_csv(
                        os.path.join(gt_pre_dir, f"{out_gene_id}.tsv.gz"),
                        sep="\t", index=False
                    )

        
        frag_df = pd.DataFrame({
            "transcript_id": tids.astype(int),
            "read_start":    sel_starts.astype(int),
            "read_end":      sel_ends.astype(int),
        })
        collapsed = (
            frag_df
            .assign(coord=frag_df["read_start"].astype(str) + "-" + frag_df["read_end"].astype(str))
            .groupby("transcript_id", sort=False)["coord"]
            .agg(",".join)
            .reset_index()
            .rename(columns={"coord": "read_coordinates"})
        )
        collapsed.to_csv(out_path, sep="\t", index=False)

    elif args.mode == "proseq":
        if "molecule_id" not in df.columns:
            raise ValueError("'molecule_id' column required for proseq mode")
        tid_to_molid = dict(zip(df["transcript_id"].astype(int), df["molecule_id"]))

        frag_df = pd.DataFrame({
            "transcript_id": tids.astype(int),
            "molecule_id":   [tid_to_molid.get(int(t), "") for t in tids],
            "read_start":    sel_starts.astype(int),
            "read_end":      sel_ends.astype(int),
        })
        frag_df.to_csv(out_path, sep="\t", index=False)

    elif args.mode == "ttseq":
        frag_df = pd.DataFrame({
            "transcript_id": tids.astype(int),
            "read_start":    sel_starts.astype(int),
            "read_end":      sel_ends.astype(int),
        })
        collapsed = (
            frag_df
            .assign(coord=frag_df["read_start"].astype(str) + "-" + frag_df["read_end"].astype(str))
            .groupby("transcript_id", sort=False)["coord"]
            .agg(",".join)
            .reset_index()
            .rename(columns={"coord": "read_coordinates"})
        )
        collapsed.to_csv(out_path, sep="\t", index=False)


if __name__ == "__main__":
    main()

